### DISCUSSION TITLE
k8s HTTP retriever securely pass api key header

### DISCUSSION THREAD

**Discussion by tomflenner** [thomaspoignant/go-feature-flag#1822] [Q&A] [answered]
Hi 👋🏻 !

We have an Azure Function which is our HTTP Retriever and the endpoint is protected by an API Key.

This API Key is stored as [sealed secret](https://github.com/bitnami-labs/sealed-secrets) in our k8s cluster.

Is it possible to reference secret in the `relayproxy.config` for api key header ?

Thanks in advance 🙏🏻

**Comment by thomaspoignant** [comment, marked as answer]
Hello @tomflenner 👋 !
I am not super familiar with how **sealed secret** works but if you can use them as environment variables you can mount your API key in the HTTP header of your HTTP retriever inside the relay proxy.

I just did a test on my local environment and you can add an environment variable that looks like `RETRIEVER_HEADERS_AUTHORIZATION=Bearer xxx` when starting your container and the header will be added to every HTTP calls made by the HTTP retriever.

Would it work for your use case?


---
EDIT: You can now override anything that is in an array thanks to this PR https://github.com/thomaspoignant/go-feature-flag/pull/1902

**Reply by tomflenner** [reply]
Hi 👋🏻 !

It seems to work for my use case.

Just for my knowledege, how this environment variable is use and read ? I assume that `RETRIEVER_...` will work If I use [retriever ](https://gofeatureflag.org/docs/relay_proxy/configure_relay_proxy#:~:text=Description-,retriever,-retriever)in the configuration but what about using multiple [retrievers](https://gofeatureflag.org/docs/relay_proxy/configure_relay_proxy#:~:text=is%20not%20set.-,retrievers,-%5B%5Dretriever) ?

Just for my training on GOF, how this configuration file works ? Does it set all as environment variable or there is a specific configuration reader in source code ?

Thanks in advance and sorry if these questions came out of scope of this thread 🙏🏻

**Reply by thomaspoignant** [reply]
> Just for my knowledege, how this environment variable is use and read ? I assume that `RETRIEVER_...` will work If I use [retriever ](https://gofeatureflag.org/docs/relay_proxy/configure_relay_proxy#:~:text=Description-,retriever,-retriever)in the configuration but what about using multiple [retrievers](https://gofeatureflag.org/docs/relay_proxy/configure_relay_proxy#:~:text=is%20not%20set.-,retrievers,-%5B%5Dretriever) ?

For `retrievers` unfortunately, it will not work for now because we are using the `koanf` library and it does not support out-of-the-box patching arrays through env variables.
This is something where we can open an issue on GOFF repo too, so we can look at that at some point _(#1841)_.

>Just for my training on GOF, how this configuration file works ? Does it set all as environment variable or there is a specific configuration reader in source code ?

As mentioned on top we are using [`knadh/koanf`](https://github.com/knadh/koanf) to handle the configuration, and we are reading several locations **pflags**, **file**, **env variables**. If you are interested in digging into that please look here: https://github.com/thomaspoignant/go-feature-flag/blob/e295dc4d0878ce1c22f6b9a6159510df4866c869/cmd/relayproxy/config/config.go#L55-L107

**Reply by tomflenner** [reply]
Thanks for all informations and for opening the issue #1841 !

Your workaround gonna be perfect for us atm since we will only use HTTP retriever (but maybe we will need to merge in future, it's YAGNI for us at this step of our POC)

I am gonna take a look into source code to see how it works !
